ml4gw/utils/slicing.py

```diff
@@ -13,6 +14,76 @@
 ]
 
 
+def unfold_windows(
+    x: torch.Tensor,
+    window_size: int,
+    stride: int,
+    drop_last: bool = True,
+):
+    """Unfold a timeseries into windows
+
+    Args:
+        x:
+            The timeseries to unfold. Can have shape
+            `(batch_size, num_channels, length * sample_rate)`,
+            `(num_channels, length * sample_rate)`, or
+            `(length * sample_rate)`
+        window_size:
+            The size of the windows to unfold from x
+        stride:
+            The stride between windows
+        drop_last:
+            If true, does not return the remainder that exists
+            when the timeseries cannot be evenly broken up into
+            windows
+
+    Returns:
+       A tensor of shape
+       `(num_windows, batch_size, num_channels, kernel_size)`,
+       `(num_windows, num_channels, kernel_size)`, or
+       `(num_windows, kernel_size)` depending on whether the
+       input tensor is 3D, 2D, or 1D
+
+       If `drop_last` is false, returns the remainder of the
+       timeseries, shaped to be compatible with the returned
+       unfolded tensor
+    """
+    reshape = list(x.shape[:-1])
+    if x.ndim == 1:
+        x = x[None, None, None, :]
+    elif x.ndim == 2:
+        x = x[None, :, None, :]
+    elif x.ndim == 3:
+        x = x[:, :, None, :]
+
+    num_windows = (x.shape[-1] - window_size) // stride + 1
+    remainder = x.shape[-1] - window_size - (num_windows - 1) * stride
+    if remainder == 0:
+        if not drop_last:
+            # create an empty tensor for consistency
+            shape = list(x.shape)
+            shape[-1] = 0
+            remainder = torch.zeros(shape, dtype=x.dtype, device=x.device)
+    else:
+        # separate x from its remainder _regardless_ of whether
+        # we end up returning it or not
+        x, remainder = torch.split(
+            x, [x.shape[-1] - remainder, remainder], dim=-1
+        )
+
+    x = unfold(x, (1, num_windows), dilation=(1, stride))
+    reshape += [num_windows, -1]
+    x = x.reshape(*reshape)
+    x = x.transpose(1, -2).transpose(0, 1)
+
+    if not drop_last:
+        reshape[-2] = 1
+        remainder = remainder.reshape(*reshape)
+        remainder = remainder.transpose(1, -2).transpose(0, 1)
```

Comment:
I think all you should need to do here is `remainder = remainder[None]` to add a dummy 1st dimension, since the rest of the dims are what we want, right?

Reply:
(or 0th dimension I guess)

Reply:
Right, yes, I forgot to move the reshape